File: server/dtypes/cregister/cregister.go
// Package cregister defines CRegister, a simple last-one-wins CRDT value.
package cregister

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/asadovsky/cdb/server/common"
	"github.com/asadovsky/cdb/server/dtypes/cvalue"
)

// CRegister is a CRDT register (last-one-wins).
// Fields are exported to support CRegister.Encode.
type CRegister struct {
	AgentId int
	Vec     *common.VersionVector
	Time    time.Time
	Val     interface{}
}

// New returns a new CRegister.
func New() *CRegister {
	return &CRegister{}
}

// DType implements CValue.DType.
func (r *CRegister) DType() string {
	return cvalue.DTypeCRegister
}

// Encode implements CValue.Encode.
func (r *CRegister) Encode() (string, error) {
	buf, err := json.Marshal(r)
	if err != nil {
		return "", err
	}
	return string(buf), nil
}

// Decode decodes the given value into a CRegister.
func Decode(s string) (*CRegister, error) {
	return nil, errors.New("not implemented")
}

func (r *CRegister) applyPatch(other *CRegister) {
	if other.Vec.After(r.Vec) || (!other.Vec.Before(r.Vec) && (other.Time.After(r.Time) || (other.Time.Equal(r.Time) && other.AgentId > r.AgentId))) {
		*r = *other
	}
}

// ApplyServerPatch implements CValue.ApplyServerPatch.
func (r *CRegister) ApplyServerPatch(patch string) error {
	// For server patches, 'patch' is an encoded CRegister.
	var other CRegister
	if err := json.Unmarshal([]byte(patch), &other); err != nil {
		return err
	}
	r.applyPatch(&other)
	return nil
}

// ApplyClientPatch implements CValue.ApplyClientPatch.
func (r *CRegister) ApplyClientPatch(agentId int, vec *common.VersionVector, t time.Time, patch string) (string, error) {
	// For client patches, 'patch' is an encoded value.
	var val interface{}
	if err := json.Unmarshal([]byte(patch), &val); err != nil {
		return "", err
	}
	other := &CRegister{
		AgentId: agentId,
		Vec:     vec,
		Time:    t,
		Val:     val,
	}
	res, err := other.Encode()
	if err != nil {
		return "", err
	}
	r.applyPatch(other)
	return res, nil
}

File: client/dtypes/cregister.js
// CRegister class.
// Mostly mirrors server/dtypes/cregister/cregister.go.

var inherits = require('inherits');

var cvalue = require('./cvalue');

////////////////////////////////////////////////////////////
// Events

inherits(Set, cvalue.Event);
function Set(isLocal, value) {
  cvalue.Event.call(this, isLocal);
  this.value = value;
}

////////////////////////////////////////////////////////////
// CRegister

inherits(CRegister, cvalue.CValue);
function CRegister(agentId, vec, time, val) {
  cvalue.CValue.call(this);
  this.agentId_ = agentId;
  this.vec_ = vec;
  this.time_ = time;
  this.val_ = val;
}

// Implements CValue.dtype.
CRegister.prototype.dtype = function() {
  return cvalue.dtypeCRegister;
};

// Decodes the given string into a CRegister.
function decode(s) {
  var x = JSON.parse(s);
  return new CRegister(x.AgentId, x.Vec, x.Time, x.Val);
}

// Implements CValue.applyPatch.
CRegister.prototype.applyPatch = function(isLocal, patch) {
  if (isLocal) {
    this.paused_ = false;
  }
  var other = decode(patch);
  // FIXME: Copy applyPatch logic from cregister.go.
  this.agentId_ = other.agentId_;
  this.vec_ = other.vec_;
  this.time_ = other.time_;
  this.val_ = other.val_;
  this.emit('set', new Set(isLocal, other.val_));
};

// Returns a native JS type that represents this value.
CRegister.prototype.get = function() {
  return this.val_;
};

// Updates this value to the given one, which must be of native JS type.
CRegister.prototype.set = function(value) {
  if (this.paused_) {
    throw new Error('paused');
  }
  this.paused_ = true;
  this.emit('patch', JSON.stringify(value));
};

////////////////////////////////////////////////////////////
// Exports

module.exports = {
  CRegister: CRegister,
  decode: decode,
  Set: Set
};
